File: 0036-valid-sudoku/0036-valid-sudoku.java
class Solution {
    public boolean isValidSudoku(char[][] board) {
       return solveSudoku(board,board[0].length); 
    }
    
    public boolean solveSudoku(char[][] board,int N) {

    for (int i = 0; i < N; i++) {
      for (int j = 0; j < N; j++) { 
       if (board[i][j] != '.' && !isValidPlacement(board, board[i][j], i, j)) return false;

        }
      }
    
    return true;
  }

  private boolean isNumInRow(char[][] board,char number,int row, int column){
    for (int i = 0; i < board.length; i++) {
      if(column != i && board[row][i]==number)return true;
    }
    return false;
  }
  private boolean isNumInColumn(char[][] board,char number, int row, int col){
    for (int i = 0; i < board.length; i++) {
      if(row != i && board[i][col]==number)return true;
    }
    return false;
  }
  private boolean isNumInBox(char[][] board,char number,int row,int col){
    int localRow=row-row%3;
    int localCol=col-col%3;

    for (int i = localRow; i < localRow+3; i++) {
      for (int j = localCol; j < localCol+3; j++) {
        if(row != i && col != j && board[i][j]==number)return true;
      }
    }
    return false;
  }
  private boolean isValidPlacement(char[][] board,char number,int row,int col){
    return !isNumInRow(board, number,row,col) &&
        !isNumInColumn( board, number,row, col)&&
      !isNumInBox(board,number,row,col);
  }
}

File: 0036-valid-sudoku/0036-valid-sudoku.py
class Solution:
    def isValidSudoku(self, board: List[List[str]]) -> bool:
        rowHa=defaultdict(set)
        colHa=defaultdict(set)
        sq=defaultdict(set)
        
        for i in range(9):
            for j in range(9):
                if board[i][j]==".":
                    continue
                if ((board[i][j] in rowHa[i]) or
                    (board[i][j] in colHa[j]) or
                    (board[i][j] in sq[(i//3,j//3)]) 
                   ):
                    return False
                rowHa[i].add(board[i][j])
                colHa[j].add(board[i][j])
                sq[(i//3,j//3)].add(board[i][j])
        return True
        
        